tests/TestCase/Database/ExpressionTypeCastingIntegrationTest.php

```diff
@@ -0,0 +1,209 @@
+                ['id' => '482b7756-8da0-419a-b21f-27da40cf8569', 'published' => 0, 'name' => 'Item 3'],
+            ]);
+
+        $query->execute();
+    }
+
+    /**
+     * Tests inserting a value that is to be converted to an expression
+     * automatically
+     *
+     * @return void
+     */
+    public function testInsert()
+    {
+        $this->_insert();
+        $query = $this->connection->newQuery()
+            ->select('id')
+            ->from('ordered_uuid_items')
+            ->order('id')
+            ->defaultTypes(['id' => 'ordered_uuid']);
+
+        $query->selectTypeMap($query->typeMap());
+        $results = $query->execute()->fetchAll('assoc');
+
+        $this->assertEquals(new UuidValue('419a8da0482b7756b21f27da40cf8569'), $results[0]['id']);
+        $this->assertEquals(new UuidValue('43e0b920481fc6d0a40d6d1740cf8569'), $results[1]['id']);
+        $this->assertEquals(new UuidValue('4c2681c048298a29a7fb413140cf8569'), $results[2]['id']);
+    }
+
+    /**
+     * Test selecting with a custom expression type using conditions
+     *
+     * @return void
+     */
+    public function testSelectWithConditions()
+    {
+        $this->_insert();
+        $result = $this->connection->newQuery()
+            ->select('id')
+            ->from('ordered_uuid_items')
+            ->where(['id' => '48298a29-81c0-4c26-a7fb-413140cf8569'], ['id' => 'ordered_uuid'])
+            ->execute()
+            ->fetchAll('assoc');
+
+        $this->assertCount(1, $result);
+        $this->assertEquals('4c2681c048298a29a7fb413140cf8569', $result[0]['id']);
+    }
+
+    /**
+     * Tests using the expression type in with an IN condition
+     *
+     * @var string
+     */
+    public function testSelectWithInCondition()
+    {
+        $this->_insert();
+        $result = $this->connection->newQuery()
+            ->select('id')
+            ->from('ordered_uuid_items')
+            ->where(
+                ['id' => ['48298a29-81c0-4c26-a7fb-413140cf8569', '482b7756-8da0-419a-b21f-27da40cf8569']],
+                ['id' => 'ordered_uuid[]']
+            )
+            ->order('id')
+            ->execute()
+            ->fetchAll('assoc');
+
+        $this->assertCount(2, $result);
+        $this->assertEquals('419a8da0482b7756b21f27da40cf8569', $result[0]['id']);
+        $this->assertEquals('419a8da0482b7756b21f27da40cf8569', $result[0]['id']);
+    }
+
+    /**
+     * Tests using an expression type in a between condition
+     *
+     * @return void
+     */
+    public function testSelectWithBetween()
+    {
+        $this->_insert();
+        $result = $this->connection->newQuery()
+            ->select('id')
+            ->from('ordered_uuid_items')
+            ->where(function ($exp) {
+                return $exp->between(
+                    'id',
+                    '482b7756-8da0-419a-b21f-27da40cf8569',
+                    '48298a29-81c0-4c26-a7fb-413140cf8569',
```

Comment:
Would not be interesting, imo. The thing I'm showing here is that passing a random uuid, can be converted to a set of expressions to remove the dashes and order it.